singularity-eos/eos/modifiers/tilted_maxwell_eos.hpp

```diff
@@ -0,0 +1,314 @@
+#include <utility>
+
+#include <ports-of-call/portability.hpp>
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/math_utils.hpp>
+#include <singularity-eos/base/root-finding-1d/root_finding.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+using namespace RootFinding1D;
+
+template <typename T>
+class TiltedMaxwellEOS : public EosBase<TiltedMaxwellEOS<T>> {
+ public:
+  // Generic functions provided by the base class. These contain
+  // e.g. the vector overloads that use the scalar versions declared
+  // here We explicitly list, rather than using the macro because we
+  // overload some methods.
+
+  // TODO(JMM): The modifier EOS's should probably call the specific
+  // sub-functions of the class they modify so that they can leverage,
+  // e.g., an especially performant or special version of these
+  using EosBase<TiltedMaxwellEOS<T>>::TemperatureFromDensityInternalEnergy;
+  using EosBase<TiltedMaxwellEOS<T>>::InternalEnergyFromDensityTemperature;
+  using EosBase<TiltedMaxwellEOS<T>>::PressureFromDensityTemperature;
+  using EosBase<TiltedMaxwellEOS<T>>::PressureFromDensityInternalEnergy;
+  using EosBase<TiltedMaxwellEOS<T>>::SpecificHeatFromDensityTemperature;
+  using EosBase<TiltedMaxwellEOS<T>>::SpecificHeatFromDensityInternalEnergy;
+  using EosBase<TiltedMaxwellEOS<T>>::BulkModulusFromDensityTemperature;
+  using EosBase<TiltedMaxwellEOS<T>>::BulkModulusFromDensityInternalEnergy;
+  using EosBase<TiltedMaxwellEOS<T>>::GruneisenParamFromDensityTemperature;
+  using EosBase<TiltedMaxwellEOS<T>>::GruneisenParamFromDensityInternalEnergy;
+  using EosBase<TiltedMaxwellEOS<T>>::PTofRE;
+  using EosBase<TiltedMaxwellEOS<T>>::FillEos;
+
+  using BaseType = T;
+
+  // give me std::format or fmt::format...
+  static std::string EosType() {
+    return std::string("TiltedMaxwellEOS<") + T::EosType() + std::string(">");
+  }
+
+  static std::string EosPyType() { return std::string("Shifted") + T::EosPyType(); }
+
+  // move semantics ensures dynamic memory comes along for the ride
+  TiltedMaxwellEOS(T &&t, Real tilt_fac) : t_(std::forward<T>(t)), tilt_(tilt_fac) {}
+  TiltedMaxwellEOS() = default;
+
+  Real g(Real u) const { return (1.0 - tilt_) * u + tilt_ * u * u;} 
+  Real gp(Real u) const { return 1.0 - tilt_ + 2 * tilt_ * u;} 
+  
+  const Real tilt_;
+  static constexpr Real tol = 1.e-10;
+  static constexpr Real eos_rho_min = 1.e-8;
+  static constexpr Real eos_rho_max = 1.e2;
+
+  auto TestMaxwellConditions(Real pin, Real temp, Real rho_p_min, Real rho_p_max) const { 
+    // Find densities above and below spinodal region with p(rho_l) = pin and p(rho_h) = pin  
+    Real rho_l, rho_h;
+    RootCounts counts_l, counts_h;
+    auto stat_l = regula_falsi([this, temp](Real rho){return this->t_.PressureFromDensityTemperature(rho, temp);}, 
+                               pin, 0.8 * rho_p_max, 
```

Comment:
why this value? And not, say `0.990 * rho_p_max`?

Reply:
In answer to most of your questions below, right now all of these values are hard coded because I was working on getting it to work for a particular EoS. I have to think a bit about what parameters we need to set in the class constructor. There are many factors that show up for initializing guesses, etc., so it would be nice if we didn't have to expose all of them and could relate them all to a few user definable parameters.

Reply:
:+1: